Add `DifferentFrom` Attribute Validator for each attribute type

chrismarget opened the pull request:
This PR adds a `DifferentFrom` attribute validator for each attribute type.

It's the result of this discussion:
https://discuss.hashicorp.com/t/framework-attribute-validation-a-differentfrom-validator/50116

The validator ensures that attributes which are set (not `null`) do not share the same value (using `attr.Value.Equal()`) as attributes at practitioner specified paths.

It's a shameless copy of the `AlsoRequires` validator with tiny logic changes.

Testing: I could't figure out how `AlsoRequires` (or any of the inter-attribute validators) are tested. I'd appreciate a pointer about that.

chrismarget-j commented:
This probably would have made more sense if I'd called it `NotEqual()` (rather than DifferentFrom()` because it throws validation failures when the attribute value's `Equal()` method comes up `true`.

bflad commented:
Hi @chrismarget 👋 Thank you for submitting this!

On our end we likely need to clarify what is permissable into this Go module or not, since otherwise this module could become infinite in scope and maintenance. This would be something like:

- Proposed validator must be generic in nature.
- Proposed validator must be a common use case.
- etc.

I believe this submission would be covered by these (currently) undocumented criteria at the moment though. We'll be taking a closer look at this on Friday.

`NotEqual()` is actually a great suggestion as you mention. In terms of unit testing, `internal/schemavalidator/also_requires_test.go` has example unit tests for that functionality which would be applicable to all the typed implementations. We'd probably be looking for something similar here. 👍

chrismarget-j commented:
Hi @bflad,

I've replaced the `DifferentFrom` nomenclature with `NotEqual` and added tests modeled on the `AlsoRequires` tests you pointed me to.

The tests have some problems because of a misunderstanding on my part.

I'd been thinking that `attr.Value`s had some cross-compatibility between, say `Int64` and `Number` types, but that doesn't seem to be the case. The various `Equal()` methods (all of them?) return `false` when an attempt to compare `attr.Value` of different types is attempted, not because of their values, but because of type assertion failure.

So, I think these test cases are invalid and should be removed.
* `compatible-types` (compares `Int64: 42` against `Number: 43`) this test case passes because the types cannot be compared by `Equal()`, not because of their different values.
* `error_compatible-types-same-value` (compares `Number: 42` against `int64: 42`) this test case expected to produce a validation error and fails because no error falls out (the matching value is not discovered by the validator).
* `incompatible-types` (compares `Int64: 42` against `String: "foo value"`) test case currently passes, but should probably be removed because such a comparison shouldn't be attempted.

I can just remove the test cases written around my bad assumption, or I can remove those test cases *and* provide some feedback to provider developers, so they know not to attempt `NotEqual` validation between attributes of different types.

Which behavior makes more sense here?
* guardrails within the `NotEqual` validator which do not permit validation attempts across types?
* no guardrails: the `NotEqual` is always content with the result of the `Equal()` method, knowing that it's likely short-circuited by type differences, even when the underlying value might be the same.

chrismarget commented:
Hi @bflad,

I've replaced the `DifferentFrom` nomenclature with `NotEqual` and added tests modeled on the `AlsoRequires` tests you pointed me to.

The tests have some problems because of a misunderstanding on my part.

I'd been thinking that `attr.Value`s had some cross-compatibility between, say `Int64` and `Number` types, but that doesn't seem to be the case. The various `Equal()` methods (all of them?) return `false` when an attempt to compare `attr.Value` of different types is attempted, not because of their values, but because of type assertion failure.

So, I think these test cases are invalid and should be removed.
* `compatible-types` (compares `Int64: 42` against `Number: 43`) this test case passes because the types cannot be compared by `Equal()`, not because of their different values.
* `error_compatible-types-same-value` (compares `Number: 42` against `int64: 42`) this test case expected to produce a validation error and fails because no error falls out (the matching value is not discovered by the validator).
* `incompatible-types` (compares `Int64: 42` against `String: "foo value"`) test case currently passes, but should probably be removed because such a comparison shouldn't be attempted.

I can just remove the test cases written around my bad assumption, or I can remove those test cases *and* provide some feedback to provider developers, so they know not to attempt `NotEqual` validation between attributes of different types.

Which behavior makes more sense here?
* guardrails: the `NotEqual` validator does not permit validation attempts across types, raises a general error, rather than a validation error.
* no guardrails: the `NotEqual` is always content with the result of the `Equal()` method, knowing that it's likely short-circuited by type differences, even when the underlying value might be the same.

chrismarget commented:
I've just opened [terraform-provider-framework #679](https://github.com/hashicorp/terraform-plugin-framework/issues/679). Clarifying that issue should answer the test question raised in my earlier comment.

chrismarget commented:
A [comment from @blfad](https://github.com/hashicorp/terraform-plugin-framework/issues/679#issuecomment-1452155210) makes it clear that testing for equality across different types isn't meaningful and should not be attempted.

I plan to remove the test cases which attempt this.

SBGoods reviewed (changes requested):
Thanks again for your contribution! 

Regarding your previous comments about the test cases comparing values of different types, I think we should opt for adding guardrails where we would throw a general error if there is a validation attempt against another type as you mentioned. I think we can keep the test cases that you mentioned but instead test for this general error to be thrown instead of a validation error.

Otherwise, I had a few comments about implementation and other test cases that could be useful. Feel free to reach out with any questions/comments.

chrismarget-j commented:
@SBGoods,

Regarding guardrails for invalid cross-type "NotEqual" validation attempts, I've used `validatordiag.InvalidAttributeTypeDiagnostic()` to produce errors like:

```
Attribute bar of type "basetypes.Int64Type", cannot be compared with attribute foo, got: basetypes.NumberType
```

I think that covers it, please feel free to nudge me in different directions regarding the construction of the error (commit imminent).

Previously you said:

> keep the test cases that you mentioned but instead check for this general error to be thrown instead of a validation error

I'm not clear if the notion of "compatible-types" in test case naming makes sense anymore. I'm leaning toward removing nomenclature and test cases which nod toward comparisons between types (other than to generate errors).

Finally, the whole validator and tests are shamelessly copied from a different validator. The test framework to handle "check for this general error to be thrown instead of a validation error" doesn't currently exist.

Before I add extra `expGenErrs int` to `testCase` and a way to tally these error types independently, I want to make sure that's what you had in mind. If there's a pre-existing pattern you'd like me to follow, please point me toward it.

chrismarget-j commented on internal/schemavalidator/not_equal.go:
```diff
@@ -0,0 +1,230 @@
+	if req.ConfigValue.IsNull() {
+		return
+	}
+
```
6f11ed2f9ced8f0db57c672a5097df7951965db7 and 8b38993a00b94625f6c59521d4813f5f07dc53d1

bflad commented:
Drive-by comment:

> Before I add extra expGenErrs int to testCase and a way to tally these error types independently, I want to make sure that's what you had in mind. If there's a pre-existing pattern you'd like me to follow, please point me toward it.

One pattern we follow upstream in terraform-plugin-framework is to use go-cmp to compare the result `diag.Diagnostics` against an expected `diag.Diagnostics` in the test case. This should allow for the full range of checking the diagnostic count, severities, summaries, details, and paths coming out. 👍 

Quick example: https://github.com/hashicorp/terraform-plugin-framework/blob/8536f49ae31ef761e3ab0a74e69fd31f4ec0fd6a/types/basetypes/float64_test.go#L32
